hyperspy/tests/test_events.py

```diff
@@ -0,0 +1,378 @@
+               3: set([f_d]), 5: set([f_c]),
+               None: set([f_a, f_b, f_c, f_d])}
+        for k, v in ref.iteritems():
+            con = self.events.a.connected(k)
+            nt.assert_equal(con, v)
+
+        con = self.events.a.connected()
+        nt.assert_equal(con, ref[None])
+
+    @nt.raises(TypeError)
+    def test_type(self):
+        self.events.a.connect('f_a')
+
+
+class TestTriggerArgResolution(EventsBase):
+
+    def setup(self):
+        self.events = he.Events()
+        self.events.a = he.Event(arguments=['A', 'B'])
+        self.events.b = he.Event(arguments=['A', 'B', 'C'])
+        self.events.c = he.Event()
+
+    def test_nargs_resolution(self):
+        self.events.a.connect(lambda x=None: nt.assert_equal(x, None), 0)
+        self.events.a.connect(lambda x: nt.assert_equal(x, 'vA'), 1)
+        self.events.a.connect(lambda x, y:
+                              nt.assert_equal((x, y), ('vA', 'vB')), 2)
+        self.events.a.connect(lambda x, A:
+                              nt.assert_equal((x, A), ('vA', 'vB')), 2)
+        self.events.a.connect(lambda x, y=None, A=None:
+                              nt.assert_equal((x, y, A),
+                                              ('vA', 'vB', None)), 2)
+        self.events.a.connect(lambda B, A:
+                              nt.assert_equal((B, A), ('vA', 'vB')), 2)
+        self.events.a.connect(lambda B, y:
+                              nt.assert_equal((B, y), ('vA', 'vB')), 2)
+        self.events.a.connect(lambda B, A=None:
+                              nt.assert_equal(B, 'vA'), 1)
+        self.events.a.connect(lambda B, y=None:
+                              nt.assert_equal(B, 'vA'), 1)
+        # Update if internal code changes, not testing this line:
+        self.events.c._connected = self.events.a._connected
+
+        self.events.a.trigger('vA', 'vB')
+        nt.assert_raises(TypeError, self.events.a.trigger, 'vA', 'vB', 'vC')
+        self.events.a.trigger(A='vA', B='vB')
+        self.events.a.trigger('vA', B='vB')
+        self.events.a.trigger(B='vB', A='vA')
+        nt.assert_raises(TypeError, self.events.a.trigger,
+                         'vA', C='vC', B='vB', D='vD')
+
+    def test_all_args_resolution(self):
+        self.events.c.connect(lambda x, y:
+                              nt.assert_equal((x, y), ('vA', 'vB')), 'all')
+        self.events.c.trigger('vA', 'vB')
+        self.events.c.connect(lambda x, y, A=None, B=None:
+                              nt.assert_equal((x, y, A, B),
+                                              ('vA', 'vB', None, None)), 'all')
+        self.events.c.trigger('vA', 'vB')
+
+        self.events.a.connect(lambda x, y, A=None, B=None:
+                              nt.assert_equal((x, y, A, B),
+                                              ('vA', 'vB', None, None)), 'all')
+        nt.assert_raises(TypeError, self.events.a.trigger, 'vA', 'vB')
```

Comment:
Event `a` has defined a fixed argument order, so the signature of  the trigger function is `a.trigger(A, B)`. This signature doesn't discriminate on how `A` and `B` were passed to it (arg or kwarg), so the parameters `A` and `B` are passed on as keyword arguments, as keywords have been given priority in the implementation. `'vA'` and `'vB'` therefore go to `A` and `B` keywords, leaving `x` and `y` unsatisfied, raising a standard python TypeError (the always confusing: `function takes at least two arguments, called with two arguments`). This would not happen for an event without this fixed argument order (event `c`), as is tested just above, as it can only take args and kwargs and pass them along as it received them.